### ISSUE TITLE
aarch64 support? 

### ISSUE THREAD

**Issue by MatthewAwesome** [spectralcode/OCTproZ#9]
Hi there! 

Really cool project!!

I attempt to compile the project in QtCreator on an Nvidia Jetson device, the Jetson Nano. 

The Jetson line of devices utilizes an aarch64 architecture as opposed to an x86. 

My guess is this key difference is why I'm having trouble compiling it. 

My guess is this comes down to the following in the cuda.pri file: 

```
unix{
	CUDA_DIR = /usr/local/cuda

	INCLUDEPATH_CUDA += /usr/include/x86_64-linux-gnu/qt5 \	#todo: is there a more general way to access the qt include directory?
		/usr/include/x86_64-linux-gnu/qt5/QtCore\
		$$CUDA_DIR/include \
		$$CUDA_DIR/samples/common/inc

	INCLUDEPATH += $$CUDA_DIR/include \
		$$CUDA_DIR/samples/common/inc

	QMAKE_LIBDIR += $$CUDA_DIR/lib64
}

```
In which case the _#todo_ could be a way of generalize the architecture, be it x86 or aarch64. 

I'll experiment with some solutions. Just thought I'd bring this to your attention! The Jetson boards are really nifty and could definitely have a place in an OCT rig, especially a lightweight one!

Thanks for being awesome! 

Matt

**Comment by spectralcode** [comment]
Hi Matt,

thanks for the feedback! 
Yes, this is great, please keep me updated. Running the project on a Jetson board would be really cool. 

Some time ago, the GitHub user @rpietruc forked the project and created a branch called "cuda9-archlinux". He changed the exact part of the code that you pointed out. You can have a look on his changes [here.](https://github.com/rpietruc/OCTproZ/commit/ab2f30dc82a5a8d28441e291f19673266fb0a286) 

Maybe this is already the solution to make it run on aarch64?

**Comment by MatthewAwesome** [comment]
Thank you for the quick reply! 

Took a look at the linked project, and implemented similar changes according the the Jetson's file system.

The project built in QtCreator!! 

However, when attempting to launch the virtual OCT system and load the sample data, the program crashed. Something about shaders and openGL libraries. 

I'll keep you posted on this front.

Perhaps there's a way to make the display portion of of the program a bit lighter weight? Or perhaps it can be modified to better suit the Jetson?

I'll look into the above and keep you posted! 

(meanwhile, since it seems like you're into some innovative OCT stuff, I'll formally correspond to your paper and we can talk collaboration!)

Thank you again!

**Comment by MatthewAwesome** [comment]
Just to keep everyone up to speed on this issue/request: 

Running into a CUDA issue. 

Namely, cudaHostRegister  is not available on ARM devices. Thus, code blocks such as lines 444-449 in cuda_code.cu (shown below) are throwing errors.
`extern "C" void cuda_registerStreamingBuffers(void* h_streamingBuffer1, void* h_streamingBuffer2, size_t bytesPerBuffer) {
	checkCudaErrors(cudaHostRegister(h_streamingBuffer1, bytesPerBuffer, cudaHostRegisterPortable));
	checkCudaErrors(cudaHostRegister(h_streamingBuffer2, bytesPerBuffer, cudaHostRegisterPortable));
	host_streamingBuffer1 = h_streamingBuffer1;
	host_streamingBuffer2 = h_streamingBuffer2;
}
` 
[This NVIDIA developer forum issue](https://forums.developer.nvidia.com/t/cudaerrornotsupported-when-calling-cv-cudahostregister-on-nvidia-tx2/60236) discusses a work around. 

In brief, it is suggested:
"_cudaHostRegister() is not supported on ARM platforms.
This is because the caching attribute of an existing allocation can’t be changed on the fly.
If required, please use cudaHostAlloc() with the flag cudaHostAllocMapped to allocate device-mapped host-accessible memory._"

I will mess around with these fixes and update accordingly.

**Comment by spectralcode** [comment]
@MatthewAwesome thank you very much for the update!
If you need to replace "cudaHostRegister" with another command, you will likely need to replace "cudaMemcpyAsync" as well (maybe it can be replaced by "cudaMemcpy")

**Comment by MatthewAwesome** [comment]
@spectralcode I've gotten the OCT data to process without error using "cudaHostAlloc". ("cudaMallocHost" also works, they two functions seem equivalent in many ways). 

The sticky point was the templating, for example, cudaHostAlloc must be called as follows: 

`checkCudaErrors(cudaHostAlloc((void**)&h_streamingBuffer2, bytesPerBuffer, cudaHostAllocPortable));`

Also, as you alluded to, cudaMemcpyAsync still functions as it should. 

All encouraging stuff! 

However, OpenGL is giving me fits as far as the 2D and 3D visualization are concerned. (1D plotting works fine)

This is the next order of business. 

Cheers!

**Comment by spectralcode** [comment]
@MatthewAwesome I was able to verify everything you reported so far and I now do understand what kind of issues you are facing with OpenGL.

The main issue here (at least for my configuration) is that Qt and the Jetson Nano are using _OpenGL ES_ and not _desktop OpenGL_. Some things (like GL_QUADS in glwindow2d.cpp) are not supported by OpenGL ES.
According the [Nvidia docs](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide%2Fsoftware_features_jetson_nano.html%23wwpID0E0BB0HA) the Jetson Nano does support OpenGL 4.5 but I don't know how to enable it in Qt.

**Comment by spectralcode** [comment]
@MatthewAwesome OCTproZ runs on the Jetson Nano! I ended up compiling Qt from source with enabled _OpenGL desktop_ option ([here are some instructions](https://github.com/spectralcode/OCTproZ/blob/master/BUILD.md#octproz-on-the-nvidia-jetson-nano-with-jetpack)) and was then able to compile and run OCTproZ after minimal modifications on the OCTproZ source code. I pushed the updated code to the main branch. 

2D and 3D volume visualization is working on the Nano. Have fun!
